### ISSUE TITLE
Use with Haunted library - Chrome - Expected redirect object or HTML element, but got: "HTMLElement".

### ISSUE THREAD

**Issue by sm-areeb** [vaadin/vaadin-router#372]
Im using vaadin with haunted and lighterhtml. Works in firefox but getting this error in chrome - 

`Error: [Vaadin.Router] Invalid route resolution result for path "/". Expected redirect object or HTML element, but got: "HTMLElement". Double check the action return value for the route.`

https://rx9o9.csb.app/

Shouldnt the behaviour be consistent? Is there something im missing?

**Comment by vlukashov** [comment]
Indeed, I would expect the same behavior from Router in all supported browsers. But right now browsers do behave differently with `haunted`-created components, and Router does not make any attempts to smooth these differences.

Apparently the `haunted` library creates components with a very special prototype chain. It [replaces](https://github.com/matthewp/haunted/blob/master/lib/component.js#L83) `HTMLElement.prototype` there with its [`Proxy`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) object, and so the components created that way technically do not pass the `instanceof HTMLElement` check which Vaadin Router relies on internally.

When it comes to proxies in the prototype chain, for arbitrary objects Firefox and Chrome behave the same:
```js
class A {}
class B extends A {}

let b1 = new B();
console.log('b1 instanceof B', b1 instanceof B); // true
console.log('b1 instanceof A', b1 instanceof A); // true

let proxy = new Proxy(A.prototype, {});
Object.setPrototypeOf(B.prototype, proxy);

let b2 = new B();
console.log('b2 instanceof B', b2 instanceof B); // true
console.log('b2 instanceof A', b2 instanceof A); // false

console.log('A.prototype.isPrototypeOf(b2)', A.prototype.isPrototypeOf(b2)); // false
```

But when the place of `A` is taken by `HTMLElement`, Firefox uses some special magic to keep the `b2 instanceof HTMLElement` check working, while Chrome does not make an exception.
```js
class B1 extends HTMLElement {}
window.customElements.define('x-b1', B1);

let b1 = new B1();
console.log('b1 instanceof B1', b1 instanceof B1); // true
console.log('b1 instanceof HTMLElement', b1 instanceof HTMLElement); // true

class B2 extends HTMLElement {}
let proxy = new Proxy(HTMLElement.prototype, {});
Object.setPrototypeOf(B2.prototype, proxy);
window.customElements.define('x-b2', B2);

let b2 = new B2();
console.log('b2 instanceof B2', b2 instanceof B2); // true
console.log('b2 instanceof HTMLElement', b2 instanceof HTMLElement); // false in Chrome, true in Firefox

console.log('HTMLElement.prototype.isPrototypeOf(b2)', HTMLElement.prototype.isPrototypeOf(b2)); // false
```

**Conclusion**
At the moment Router does not support the `haunted` library (though it happens to work in Firefox). It's possible to add a bit of code into Router to handle this special case.

One option to try would be replacing [`if (result instanceof HTMLElement)`](https://github.com/vaadin/vaadin-router/blob/master/src/router.js#L528) with (`if (result instanceof HTMLElement) || (result[Symbol.toStringTag] === 'HTMLElement')`. A PR with this change and tests to cover it would be welcome.

**Comment by matthewp** [comment]
Fixing this in Haunted. Thanks for the heads up!

**Comment by matthewp** [comment]
Fixed in Haunted 4.5.3, thanks so much for doing the research on this one. Pretty surprising to me!  https://github.com/matthewp/haunted/releases/tag/v4.5.3

**Comment by vlukashov** [comment]
Thanks for a quick fix, @matthewp!
@sm-areeb, please try using the newly released haunted v4.5.3. That should fix the issue for you.

**Comment by sm-areeb** [comment]
Thanks @vlukashov, @matthewp. It works now
